GOSSIP.md

```diff
@@ -34,45 +35,252 @@ also used to determine which node should keep its initiated connection open in
+maximum capacity of 16384 peers.
+
+To prevent attackers from predicting how good peers are evicted and replace them
+by compromised peers, the eviction is done per-bucket; the buckets are selected
+from the peer address and a secret to randomize the selection process.
+
+When a peer is verified the first time by connecting to it using the Noise
+protocol (only outgoing connections), it is removed from the unverified pool
+and the steps to add it to the verified pool are:
+
+1. A subset of 16 buckets are selected based on the secret and the IP of the
+peer to be added.
+2. From this subset single bucket is selected based on the secret and the port
+of the peer to be added; the IP and port are different discriminators to prevent
+peers sharing the same port to affect the whole pool.
+3. If the bucket is full, one peer has to be evicted. This is done by first
+cleaning the peers that weren't gossiped for a long time, and if the bucket is
+still full, by selecting a random one with a bias toward the ones that are not
+connected and the last connection was the longest time ago. Trusted peers are
+never evicted.
+4. The eventually evicted peer is added to the unverified pool; its retry
+counter and last retry time are reset. If the node is connected to the evicted
+peer, the connection is closed.
+
+See [Bucket Selection](#bucket-selection) for more details on how the buckets
+are selected from the secret and other discriminators.
+
+#### Technical Details (WIP)
+
+##### Constants
+
+Feedback on the constant selection?
+
+##### Bucket Selection
+
+The selection of a subset from a list of buckets could be done in various
+ways. The current idea is to hash the secret combined with the discriminator and
+extract groups of bits from the resulting hash as indexes in the list.
+
+I am not sure of cryptographic validity of this solution; this probably needs
+a bit more research.
+
+##### (optional) Peer Duplication in Unverified Pool
+
+The duplication of peers in the unverified pool to favor selection of frequently
+seen peers is marked as optional because I am not sure it is worth doing.
+
+#### Known limitations
+
+With the current constants:
+
+- The maximum cumulated number of peers in the pools is from 24576 to 81920
+depending on peer duplication in the unverified pool.
+- The maximum number of connected peers is 16384; it will probably be lower due
+to the clusterization and eviction algorithm.
+- The maximum number of peers that can be added through gossip by peers sharing
+the same IP is from 512 to 4016 depending on peer duplication in the unverified
+pool.
+- The maximum number of neighboring peers sharing the same IP a single node
+can add through gossip is from 128 to 1024 depending on peer duplication in the
+unverified pool.
+
+#### Concerns (WIP)
+
+- There is always a cost for updating an unverified gossiped peer:
+
+  If a peer is not in the verified pool BUT already was in the unverified pool,
```

Comment:
I think this is a problem especially in the beginning where the total set of peers/nodes is pretty small...

I might be missing something, but the fact that we are gossiped the same peer from multiple sources should favor this peer? If we simply remove it and then add it again that effect is missing?

Reply:
The issue with not cleaning the references first is that we don't control duplication, we could end up with 1024 (one per bucket) references, this would go against the idea of limiting pool poisoning. Another option is to completely forget about duplication and keep the peer in the bucket it was added the first time.